src/BoundsInference.cpp

```diff
@@ -109,27 +110,120 @@ class BoundsInference : public IRMutator {
+            }
+            if (is_update) {
+                for (const Expr &arg : def.args()) {
+                    if (!predicates.empty()) {
+                        Stmt cond_arg;
+                        for (const Expr &pred : predicates) {
+                            cond_arg = Evaluate::make(Call::make(Int(32), "dummy", {arg}, Call::PureIntrinsic));
+                            cond_arg = IfThenElse::make(likely(pred), cond_arg);
+                        }
+                        result.push_back(CondValue(const_true(), cond_arg));
+                    } else {
+                        result.push_back(CondValue(const_true(), arg));
+                    }
+                }
+            }
+
+            const vector<Specialization> &specializations = def.specializations();
+            for (size_t i = specializations.size(); i > 0; i--) {
+                Expr s_cond = specializations[i-1].condition;
+                const Definition &s_def = specializations[i-1].definition;
+
+                // Else case (i.e. specialization condition is false)
+                for (auto &i : result) {
+                    i.cond = simplify(!s_cond && i.cond);
+                }
+
+                // Then case (i.e. specialization condition is true)
+                vector<CondValue> s_result = compute_exprs_helper(s_def, is_update);
+                for (auto &i : s_result) {
+                    i.cond = simplify(s_cond && i.cond);
+                }
+                result.insert(result.end(), s_result.begin(), s_result.end());
+            }
+
+            return result;
+        }
+
+        // Computed expressions on the left and right-hand sides. This also
+        // pushes all reduction domains it encounters into the 'rdoms' set
+        // for later use.
         void compute_exprs() {
-            if (stage == 0) {
-                exprs = func.values();
+            bool is_update = (stage != 0);
+            if (!is_update) {
+                exprs = compute_exprs_helper(func.definition(), is_update);
             } else {
-                const UpdateDefinition &r = func.updates()[stage - 1];
-                exprs = r.values;
-                exprs.insert(exprs.end(), r.args.begin(), r.args.end());
-                if (r.domain.defined()) {
-                    exprs.push_back(r.domain.predicate());
+                const Definition &def = func.update(stage - 1);
+                exprs = compute_exprs_helper(def, is_update);
+            }
+        }
+
+        // Check if the dimension at index 'dim_idx' is always pure (i.e. equal to 'dim')
+        // in the definition (including in its specializations)
+        bool is_dim_always_pure(const Definition &def, const string& dim, int dim_idx) {
+            const Variable *var = def.args()[dim_idx].as<Variable>();
+            if ((!var) || (var->name != dim)) {
+                return false;
+            }
+
+            for (const auto &s : def.specializations()) {
```

Comment:
I'm not sure what the current Halide convention is on this, but I'd prefer a type here instead of auto. I think auto is only good when the type is immediately obvious. Here, I'm pretty sure it's 'Definition', but I can't be certain.